### ISSUE TITLE
UnicodeEncodeError on localised response.reason message (bad use of IOError)

### ISSUE THREAD

**Issue by thomasnoel** [psf/requests#4489]
## Expected result

Python 2.7, requests 

A website returns 'Non-Autoris\xe9' (non authorized, in French) as response.reason on a 401 error. Before #3554 (https://github.com/requests/requests/pull/3554/files) a response.raise_for_error() returns this : 
```python
import requests
resp = requests.get('http://www.example.net/admin/')  # a website returning « 401 Non Autorisé »
try:
    resp.raise_for_error()
except Exception as e:
    pass
print e
```
401 Client Error: Non-Autoris for url: http://www.example.net/admin/

(\xe9 is just removed, not really an issue)

## Actual Result

After #3554 : 

UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 29: ordinal not in range(128)

## Why ?

It's because HTTPError is in fact an IOError : the first init argument can't be a unicode string. See https://docs.python.org/2/library/exceptions.html#exceptions.EnvironmentError

```python
>>> unicode(IOError(12, u'é'))
u'[Errno 12] \xe9'
>>> unicode(IOError(u'é'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 0: ordinal not in range(128)
>>> unicode(Exception(u'é'))
u'\xe9'
```

**Comment by sigmavirus24** [comment]
Sadly we can't change the exception inheritance order at this point. I wonder if it makes sense to `encode` to ascii and to ignore any errors doing so. I wonder if `IOError` is okay with `unicode` on python 3 though.

**Comment by bdauvergne** [comment]
It is:
```
>>> str(IOError(u'é'))
'é'
```

**Comment by nateprewitt** [comment]
Yep, this was an instance of a younger, dumber me not fully rooting out the cause of the issue. Our workaround for this in the tests was to access `e.value.args[0]` which won't throw the ascii exception (yeah, I don't know why that was deemed ok either).

So from the looks of it, we can still support this for Python 3 but it's never really worked for Python 2, at least not with `raise_for_status`. I think @sigmavirus24 is right that we should go back to ignoring unrecognized values for Python 2.

**Comment by sethmlarson** [comment]
In an effort to clean up the issue tracker to only have issues that are still relevant to the project we've done a quick pass and decided this issue may no longer be relevant for a variety of potential reasons, including:

* Applies to a much older version, unclear whether the issue still applies.
* Change requires a backwards incompatible release and it's unclear if the benefits are worth the migration effort from the community.
* There isn't a clear demand from the community on the change landing in Requests.

If you think the issue should remain open, please comment so below or open a new issue and link back to the original issue. Again, thank you for opening the issue and for the discussion, it's much appreciated.
